concordium-node/src/bin/cli.rs

```diff
@@ -485,36 +485,70 @@ async fn maybe_do_out_of_band_catchup(
-        match from {
-            Ok(path) => {
-                info!("Starting out of band catch-up");
-                if let Err(e) = consensus.import_blocks(&path) {
-                    if import_stopped.load(atomic::Ordering::Acquire) {
-                        info!("Out of band catchup stopped.");
-                    } else {
-                        info!("Could not complete out of band catch-up due to: {:#}.", e);
-                    }
-                } else {
-                    info!("Completed out of band catch-up from {}.", download_url);
-                }
-                // attempt to properly clean up the downloaded file.
-                if let Err(e) = path.close() {
-                    error!("Could not delete the downloaded file: {}", e);
-                }
-            }
-            Err(e) => {
-                // report error, but continue startup
-                error!("Downloading catch-up files failed: {}.", e)
+        info!("Starting out of band catch-up");
+        if let Err(e) = import_missing_blocks(consensus, download_url, data_dir_path).await {
+            if import_stopped.load(atomic::Ordering::Acquire) {
+                info!("Out of band catchup stopped.");
+            } else {
+                info!("Could not complete out of band catch-up due to: {:#}.", e);
             }
+        } else {
+            info!("Completed out of band catch-up from {}.", download_url)
         }
     }
 }
 
-async fn download_catchup_files(
-    download_url: url::Url,
+#[derive(serde::Deserialize)]
+struct BlockChunkData {
+    filename:           String,
+    genesis_index:      u64,
+    first_block_height: u64,
+    last_block_height:  u64,
+}
+
+async fn import_missing_blocks(
+    consensus: &ConsensusContainer,
+    index_url: &url::Url,
     data_dir_path: &Path,
-) -> anyhow::Result<TempPath> {
+) -> anyhow::Result<()> {
+    let json_consensus_value: serde_json::Value =
+        serde_json::from_str(&consensus.get_consensus_status())?;
+    let best_block_height = json_consensus_value["bestBlockHeight"].as_u64().unwrap();
+    let current_genesis_index = json_consensus_value["genesisIndex"].as_u64().unwrap();
+    let current_genesis_block_hash =
+        json_consensus_value["currentEraGenesisBlock"].as_str().unwrap();
+    let json_genesis_block_value: serde_json::Value =
+        serde_json::from_str(&consensus.get_block_info(current_genesis_block_hash)?)?;
+    let current_genesis_block_height = json_genesis_block_value["blockHeight"].as_u64().unwrap();
+    let index_reader = reqwest::get(index_url.clone())
+        .await?
+        .bytes_stream()
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
+        .into_async_read();
+    let mut chunk_records = csv_async::AsyncReaderBuilder::new()
+        .comment(Some(b'#'))
+        .create_deserializer(index_reader)
+        .into_deserialize();
+    while let Some(result) = chunk_records.next().await {
```

Comment:
There does not seem to be any check that the genesis hash in the file matches ours. We should log an appropriate error message in that case, so the user can correct the issue.

Reply:
Since the (re)genesis hashes are stored in the index files as comments, I had to stream the index file twice, the first time to read the comments, the second time to parse the CSV rows.

Perhaps, it's not ideal because I could separate the comments from CSV rows while reading the download stream only once, but code readability would suffer, in my opinion. Also, the first stream is, in fact, only used to read the very first genesis hash comment, so the whole index file won't be downloaded twice. Almost certainly it won't matter in practice, because the index file will be much smaller than even one block chunk.

An alternative to storing the genesis hashes as comments would be to make them fields, but that would unnecessarily blow up the index file size (adding 64 ASCII characters needed to represent a 256-bit hash to each row will more than double the row length).